### ISSUE TITLE
Training and testing steps

### ISSUE THREAD

**Issue by jamie212** [NcolasDran/sketch-colorization-based-on-reference-structure#1]
Can you provide the steps for training and testing, please?

**Comment by NcolasDran** [comment]
Hi,  
To run the code, the **`main.py`** script must be executed, this file reads the information from **`./configs/training.cfg`** that contains instructions for training or testing. Specifically, the configuration file is divided into three sections: [Commons], [Training], and [Testing].

- In **[Commons]** there are the keys `train_model` and `test_model`, which must be assigned values True or False, depending on whether you want to train or test.
- In **[Training]**, the most important keys are `dataset_name`, `split_img`, `fixed_size_imgs` and `fine_tune`.
	+ The key `dataset_name` indicates the name of the folder where the images are located (within the data folder). The training dataset of the original paper is the [Anime Sketch Colorization Pair](https://www.kaggle.com/datasets/ktaebum/anime-sketch-colorization-pair), so the folder must be *'anime_sketch_colorization_pair'*.
	+ The `split_img` key indicates whether the images in the dataset should be split. This key exists because each image of the original training dataset (for the Main Model) is a pair of sketch and colored sketch.
	+ The key `fixed_size_imgs` must be True if all images have the same size, and this is the case for the training dataset for the Main Model. On the other hand, the images to train the Fine-Tune Model have different sizes, so it should be left as False, this adds padding to make square images.
	+ The `fine_tune` key indicates the type of architecture the network will have. Set the value to False to train the Main Model (from the original paper), and set the value to True to train the Fine-Tune Model (using the modifications I added). As a reminder, the Main Model uses one dataset and the Fine-Tune Model uses another.
- In **[Testing]** there are the same keys, which indicate the same thing, but applied to the testing context.

To make the execution easier, I updated the values in `training.cfg` to train the Main Model. And also added an **Execution** section in the README.

**Comment by jamie212** [comment]
Thank you so much for the clarification! 
So, are you suggesting that the main model was trained on the Anime Sketch Colorization Pair dataset, 
and the One Punch Man Manga Colorization dataset was utilized for training the fine-tuned model?

**Comment by NcolasDran** [comment]
Yes, that's right.  
The main model is the same model proposed by the original [paper](https://link.springer.com/article/10.1007/s41095-021-0228-6), and was trained on the Anime Sketch Colorization Pair dataset.

Once that model was trained, I wanted to improve it or adapt it to other images, so I created the One Punch Man Manga Colorization dataset and trained the main model with these new images (using the model weights from the first training). This new model I called *'fine-tuned model'*, and has some small modifications in the architecture, that I explain in the README. So, to clarify, in order to obtain the results of the fine-tuned model it is necessary to train the main model first.

**Comment by sjdfhw** [comment]
Hello, first of all this model training need to use the official data set in the JSON file? Second, I added some code to save the test-generated result image after coloring. But I trained a lot of epoch, and the test results pictures for each epoch were not colored correctly. Is it because I added the wrong code? If I want to see what the finished image looks like, how do I change the code? Looking forward to your reply！

**Comment by jamie212** [comment]
> Yes, that's right. The main model is the same model proposed by the original [paper](https://link.springer.com/article/10.1007/s41095-021-0228-6), and was trained on the Anime Sketch Colorization Pair dataset.
> 
> Once that model was trained, I wanted to improve it or adapt it to other images, so I created the One Punch Man Manga Colorization dataset and trained the main model with these new images (using the model weights from the first training). This new model I called _'fine-tuned model'_, and has some small modifications in the architecture, that I explain in the README. So, to clarify, in order to obtain the results of the fine-tuned model it is necessary to train the main model first.

Apologies for the inconvenience. I am inquiring about the One Punch Man Manga Colorization dataset you shared via the Google link. When I attempt to download it, the size shows as over 4GB, but I encounter a failure during the extraction process. I've noticed that a single image in the test collection is an unusually large 127GB, which leads me to wonder if there has been an error? Or is this an expected size for the image?

**Comment by NcolasDran** [comment]
> Hello, first of all this model training need to use the official data set in the JSON file? Second, I added some code to save the test-generated result image after coloring. But I trained a lot of epoch, and the test results pictures for each epoch were not colored correctly. Is it because I added the wrong code? If I want to see what the finished image looks like, how do I change the code? Looking forward to your reply！

Hi @sjdfhw ,  
For the first point, from the official dataset (Anime Sketch Colorization Pair) the images from the *'train'* and *'val'* folders are used, the JSON files in the *'cologram'* folder are not used.

Regarding the second point, I assume that you are saving the images generated directly by the model, specifically, the `out_img = net.colorization_network(...)` variable in `./training/testing.py`. However, the images generated by the model are normalized, this is because to better train the model the training images are normalized, which is why the inverse transformation must be performed to obtain the real images. This transformation is already in the code from the testing file, in the line `inv_out_img = INVERSE_NORMALIZE(out_img)`, the ***inv_out_img*** variable has the finished images.

**Comment by NcolasDran** [comment]
> > Yes, that's right. The main model is the same model proposed by the original [paper](https://link.springer.com/article/10.1007/s41095-021-0228-6), and was trained on the Anime Sketch Colorization Pair dataset.
> > Once that model was trained, I wanted to improve it or adapt it to other images, so I created the One Punch Man Manga Colorization dataset and trained the main model with these new images (using the model weights from the first training). This new model I called _'fine-tuned model'_, and has some small modifications in the architecture, that I explain in the README. So, to clarify, in order to obtain the results of the fine-tuned model it is necessary to train the main model first.
> 
> Apologies for the inconvenience. I am inquiring about the One Punch Man Manga Colorization dataset you shared via the Google link. When I attempt to download it, the size shows as over 4GB, but I encounter a failure during the extraction process. I've noticed that a single image in the test collection is an unusually large 127GB, which leads me to wonder if there has been an error? Or is this an expected size for the image?

@jamie212 I just downloaded the file from the same link, and I had no problems during the extraction process (it takes about 40 seconds for me). The specific size of the subfolders is **3.3 GB** for train, **619.3 MB** for test and **290.4 MB** for val.

I don't know what the problem could be, maybe the extraction method? I use WinRAR to extract the file. If the problem continues, I could create another file with the images to see if that fixes it.

**Comment by jamie212** [comment]
> > > Yes, that's right. The main model is the same model proposed by the original [paper](https://link.springer.com/article/10.1007/s41095-021-0228-6), and was trained on the Anime Sketch Colorization Pair dataset.
> > > Once that model was trained, I wanted to improve it or adapt it to other images, so I created the One Punch Man Manga Colorization dataset and trained the main model with these new images (using the model weights from the first training). This new model I called _'fine-tuned model'_, and has some small modifications in the architecture, that I explain in the README. So, to clarify, in order to obtain the results of the fine-tuned model it is necessary to train the main model first.
> > 
> > 
> > Apologies for the inconvenience. I am inquiring about the One Punch Man Manga Colorization dataset you shared via the Google link. When I attempt to download it, the size shows as over 4GB, but I encounter a failure during the extraction process. I've noticed that a single image in the test collection is an unusually large 127GB, which leads me to wonder if there has been an error? Or is this an expected size for the image?
> 
> @jamie212 I just downloaded the file from the same link, and I had no problems during the extraction process (it takes about 40 seconds for me). The specific size of the subfolders is **3.3 GB** for train, **619.3 MB** for test and **290.4 MB** for val.
> 
> I don't know what the problem could be, maybe the extraction method? I use WinRAR to extract the file. If the problem continues, I could create another file with the images to see if that fixes it.

It seems that this was indeed the issue; I switched to WinRAR and it worked. Thank you very much!

**Comment by sjdfhw** [comment]
> > 你好，首先这个模型训练需要用到JSON文件中的官方数据集吗？其次，我添加了一些代码来保存着色后测试生成的结果图像。但是我训练了很多 epoch，每个 epoch 的测试结果图片都没有正确着色。是因为我加错了代码吗？如果我想查看完成的图像是什么样子的，如何更改代码？期待您的回复！
> 
> 嗨，对于第一点，来自官方数据集（Anime Sketch Colorization Pair）使用了“train”和“_val_”文件夹中的图像，不使用_“__cologram”_文件夹中的JSON文件。
> 
> 关于第二点，我假设您正在保存模型直接生成的图像，特别是 中的变量。但是，模型生成的图像是归一化的，这是因为为了更好地训练模型，训练图像是归一化的，这就是为什么必须进行逆变换才能获得真实图像的原因。此转换已在测试文件的代码中，在行中，_**inv_out_img**_变量具有完成的图像。`out_img = net.colorization_network(...)``./training/testing.py``inv_out_img = INVERSE_NORMALIZE(out_img)`

Thank you very much, I have got the result of the test picture, “inv_color_sketch” can perfectly restore the reference picture, but “inv_out_img” generated only shape information, no color information. Second, I found that the loss of information during the training model has been particularly large, there is no downward trend, want to know how to solve? The server terminal displays the following information:
Start: 2023-11-06 09:11:17
Running epoch... 


================================
2023-11-06 09:15:39 | Epoch 67

rec_loss: 197.004
adv_loss: 23766.821
total_loss: 23963.825
================================
 --- 261 seconds total ---


 --------------------
|     VALIDATION     |
 --------------------

Running validation... 

WARNING: default feature extractor (InceptionNet V2) is used.
WARNING: default feature extractor (InceptionNet V2) is used.
Looking forward to your reply！

**Comment by sjdfhw** [comment]
> > Hello, first of all this model training need to use the official data set in the JSON file? Second, I added some code to save the test-generated result image after coloring. But I trained a lot of epoch, and the test results pictures for each epoch were not colored correctly. Is it because I added the wrong code? If I want to see what the finished image looks like, how do I change the code? Looking forward to your reply！
> 
> Hi @sjdfhw , For the first point, from the official dataset (Anime Sketch Colorization Pair) the images from the _'train'_ and _'val'_ folders are used, the JSON files in the _'cologram'_ folder are not used.
> 
> Regarding the second point, I assume that you are saving the images generated directly by the model, specifically, the `out_img = net.colorization_network(...)` variable in `./training/testing.py`. However, the images generated by the model are normalized, this is because to better train the model the training images are normalized, which is why the inverse transformation must be performed to obtain the real images. This transformation is already in the code from the testing file, in the line `inv_out_img = INVERSE_NORMALIZE(out_img)`, the _**inv_out_img**_ variable has the finished images.

Thank you very much, I have got the result of the test picture, “inv_color_sketch” can perfectly restore the reference picture, but “inv_out_img” generated only shape information, no color information. Second, I found that the loss of information during the training model has been particularly large, there is no downward trend, want to know how to solve? The server terminal displays the following information:
Start: 2023-11-06 09:11:17
Running epoch... 


================================
2023-11-06 09:15:39 | Epoch 67

================================
 --- 261 seconds total ---

rec_loss: 197.004
adv_loss: 23766.821
total_loss: 23963.825
 --------------------
|     VALIDATION     |
 --------------------

Running validation... 

WARNING: default feature extractor (InceptionNet V2) is used.
WARNING: default feature extractor (InceptionNet V2) is used.
Looking forward to your reply！

**Comment by sjdfhw** [comment]
> > Hello, first of all this model training need to use the official data set in the JSON file? Second, I added some code to save the test-generated result image after coloring. But I trained a lot of epoch, and the test results pictures for each epoch were not colored correctly. Is it because I added the wrong code? If I want to see what the finished image looks like, how do I change the code? Looking forward to your reply！
> 
> Hi @sjdfhw , For the first point, from the official dataset (Anime Sketch Colorization Pair) the images from the _'train'_ and _'val'_ folders are used, the JSON files in the _'cologram'_ folder are not used.
> 
> Regarding the second point, I assume that you are saving the images generated directly by the model, specifically, the `out_img = net.colorization_network(...)` variable in `./training/testing.py`. However, the images generated by the model are normalized, this is because to better train the model the training images are normalized, which is why the inverse transformation must be performed to obtain the real images. This transformation is already in the code from the testing file, in the line `inv_out_img = INVERSE_NORMALIZE(out_img)`, the _**inv_out_img**_ variable has the finished images.

@NcolasDran 
Thank you very much, I have got the result of the test picture, “inv_color_sketch” can perfectly restore the reference picture, but “inv_out_img” generated only shape information, no color information. Second, I found that the loss of information during the training model has been particularly large, there is no downward trend, want to know how to solve? The server terminal displays the following information:
2023-11-06 09:11:17 | Epoch 66

PSNR: 6.816
SSIM: 0.434
FID: 359.971
================================
 --- 703 seconds total ---


 Creating checkpoint...
 Checkpoint created in 'data/checkpoints'


 -------------
|   Epoch 67   |
 -------------

Start: 2023-11-06 09:11:17
Running epoch... 


================================
2023-11-06 09:15:39 | Epoch 67

rec_loss: 197.004
adv_loss: 23766.821
total_loss: 23963.825
================================
 --- 261 seconds total ---


 --------------------
|     VALIDATION     |
 --------------------

Running validation... 

WARNING: default feature extractor (InceptionNet V2) is used.
WARNING: default feature extractor (InceptionNet V2) is used.
Looking forward to your reply！

**Comment by NcolasDran** [comment]
> @NcolasDran Thank you very much, I have got the result of the test picture, “inv_color_sketch” can perfectly restore the reference picture, but “inv_out_img” generated only shape information, no color information. Second, I found that the loss of information during the training model has been particularly large, there is no downward trend, want to know how to solve? The server terminal displays the following information:

*'inv_color_sketch'* is the ground-truth image and *'inv_out_img'* is the predicted image, and the prediction should be as close as possible to the ground-truth. If the generated prediction only show the shape information and the training loss value is high, then there is probably an error in the training. I think you may have different values in `./configs/training.cfg`, please check the configuration as indicated in the **Execution** section in the **README**.

**Comment by NcolasDran** [comment]
Regarding the evaluation phases (validation and testing), the WARNING messages displayed on the terminal come from the PIQ library, so there is nothing I can do unfortunately.  
But I also realized that I had a mistake in the code, the validation phase should only be performed with the *'opm_colorization_dataset'*. The *'anime_sketch_colorization_pair'* dataset doesn't have validation phase because it only has 'train' and 'val' images, and the 'val' images are used for the testing phase (as done in the original paper). I apologize for this error and I have already updated the code with the changes.  
So please download the new changes to have the updated version @sjdfhw @jamie212 .
